runtime/caml/mlvalues.h

```diff
@@ -226,6 +230,10 @@ where 0 <= R <= 31 is HEADER_RESERVED_BITS, set with the
 /* Fields are numbered from 0. */
 #define Field(x, i) (((volatile value *)(x)) [i]) /* Also an l-value. */
 
+#define Field_atomic(x, i) \
+  atomic_load_explicit(Op_atomic_val((volatile value *)(x) + i), \
+                       memory_order_relaxed)
```

Comment:
If this is a problem for tsan then this will also likely affect FFI code (legacy and new) written by users. Does this mean that adding `volatile` to `Field` was not enough of a fix in OCaml 5 because this will create false positives for tsan?
The original idea was that a volatile read would behave closely-enough to a relaxed read: is there by chance an option to convince tsan of that?



Also, if this is really needed, then some time ago it was discussed that it would be beneficial to split the Field macro in two (one l-value and one r-value). This suggest to simply call this macro Read_field or Field_load or something like that.

Reply:
Actually, I think the `volatile` is not really needed here as we only need to tell the compiler that the field read is atomic. TSan does not see that the type is volatile so it won’t make a difference on its side.

Yes, maybe `Field_atomic_load` would be better to make explicit that it is never an l-value.

Reply:
I failed to get my point across. What I am worrying about is that if you need to replace `Field` with `Field_atomic` to avoid a warning from tsan, then maybe one also needs to do the replacement all of legacy and new C FFI code written by users. Am I missing something?

Reply:
There seems to be support for volatile instrumentation in both gcc and clang TSan implementations using the `tsan-distinguish-volatile`. See [there](https://lore.kernel.org/all/20200521142047.169334-4-elver@google.com/) for how this is used in the Linux kernel, which also uses volatile as part of its memory model (this was the inspiration for @gadmm's suggestion to use volatile in the OCaml runtime I believe).

Reply:
In other words: you may not need to change anything here if you use the same approach as the Linux kernel of generating versions of volatile reads/writes that explicitly silence TSan.

Reply:
Note that you will still have false positive for dependency ordering, see e.g. https://github.com/google/sanitizers/issues/882

Reply:
I see, enabling this option does sound useful, if non-trivial (apparently one must implement the instrumentation functions for `volatile` oneself, and the behavior differs quite a lot between GCC and Clang). We’ll see whether it allows to remove warnings on the runtime.

However, the use of `volatile` is still an ersatz of real atomic accesses intended for legacy code. I think that we should strive to use atomic accesses in the runtime where necessary, provided that it does not require too many changes.

Reply:
I don't agree with your reasoning. It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use `volatile` to enforce our memory model, for performance reasons. We collectively agreed on that.

Now you propose to deviate from this specific in a few places to silence TSan warnings. I think this is the wrong approach. My thinking:
1. Either we believe that the "volatile" model that we agreed on is the right design, and we should use tools that understand it. (If the Linux kernel uses the same model, then presumably those tools exist *or* there are other people interesting in making them exist.)
2. Or we believe that being able to use TSan unmodified is a strong argument in favor of moving to a different model (using 2A instead of 2B in the classification of @gadmm in https://github.com/ocaml/ocaml/issues/10992#issuecomment-1104033399), and we should discuss moving to a new model *in general*, instead of just in a couple places that currently raise warnings.

Reply:
> It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use volatile to enforce our memory model, for performance reasons.

To clarify, using `volatile` was more of a stop-gap measure. #10992 is still an open issue with two proposed paths, with little to no discussion happening. Since then, in private conversations with one memory-model expert we agreed that one solution does not work, whereas with another memory-model expert we agreed that the other solution would not work. (I prefer to clarify that I do not consider myself an expert on memory models and more of an observer as far as #10992 is concerned.)

In particular if you instrument volatile accesses as relaxed reads for TSan, you will still have a lot of false positives (for the publication stuff), whereas if you instrument it as an acquire read for TSan, then I suspect you might have a lot of false negatives. Either way it is already good to have experience from trying to use TSan as this might affect how we think about the memory-model problem.

Reply:
@OlivierNicole Regarding initializing writes, did you manage to not instrument the ones done from C code?

Reply:
> I don't agree with your reasoning. It took a lot of discussion to decide on which style we wanted for our C code, and https://github.com/ocaml/ocaml/issues/10992 proposed to use volatile to enforce our memory model, for performance reasons. We collectively agreed on that.

I can’t find those performance reasons you mention. Could you point me to them? In the linked issue `volatile` is clearly advocated as a stop-gap measure for existing code, as @gadmm said.

Reply:
My understanding (but I'm not the expert in this discussion thread):
- for runtime data structures that are only used from C code, we should use C11 atomics 
- but for the OCaml heap and any other data that is accessed both from the C runtime and from the OCaml mutator, we currently use `(volatile *)` following the Linux mode:
  + using C11 atomics does not provide any correctness guarantees in presence of races from OCaml accesses, we need to reason on the assembly level anyway
  + using `relaxed` may in general not be correct (we typically don´t want to have the reads hoisted out of loops, etc.)
  + using `consume` or `acquire` or `sequential` may be too expensive

Reply:
Thank you, I had missed the fact that `relaxed` may not be correct.

Reply:
There are two aspects.
- Why do we use volatile instead of relaxed atomic -> because of backwards compatibility.
- Why do we use relaxed atomic or volatile instead of an acquire load (and cross our collective fingers regarding dependency ordering) -> one reason can be performance (see the bad performance of the SRA model in the memory model paper).

I think that there is no good reason to break compatibility if not for going all the way to a standards-compliant solution (i.e. acquire).

One model at #10992 advocated in favour of strict C11 compliance at the expense of backwards-compatibility (which would only be preserved for single-threaded C FFI code or code synchronized separately, e.g. with a mutex). It wants to add a `Load_field` macro that does an acquire load but is only strictly necessary for mutable fields (of non-immediate types, probably).

The other solution is to try to make sense of the current usage of dependency ordering in terms of an extension of the C11 memory model, and working together with people who would like to have some form of dependency ordering in C to add it to the standard.

The latter seems a bit more far-fetched, but from private discussions, there might be a possible path (though we probably want more than a theoretical possibility, we want people actually working on it). Regarding the former, some devs seemed to agree in private that asking people to move from `Field` to `Load_field` with no observable benefit is doomed to fail—there will always remain incorrect uses of `Field` in the wild and it might even be the majority. Though it would have the benefit for OCaml devs of shifting responsibility onto programmers.

In the meanwhile, using `volatile` is indeed a stop-gap measure, we agree it is not good, and there may be some conclusions to draw at some point from the unwillingness or inability to discuss the issue at #10992. That's why any input to continue the discussion at #10992, in particular experience from implementing the data-race detector, is welcome!